From stable-bounces@linux.kernel.org  Thu Jul 14 16:54:10 2005
To: stable@kernel.org
From: akpm@osdl.org
Date: Thu, 14 Jul 2005 16:46:26 -0700
Cc: akpm@osdl.org, mostrows@watson.ibm.com
Subject: [patch] rocket.c: Fix ldisc ref count handling

From: Michal Ostrowski <mostrows@watson.ibm.com>

If bailing out because there is nothing to receive in rp_do_receive(),
tty_ldisc_deref is not called.  Failure to do so increases the ref count=20
and causes release_dev() to hang since it can't get the ref count to 0.

Signed-off-by: Michal Ostrowski <mostrows@watson.ibm.com>
Signed-off-by: Andrew Morton <akpm@osdl.org>
Signed-off-by: Chris Wright <chrisw@osdl.org>
---

 drivers/char/rocket.c |    3 ++-
 1 files changed, 2 insertions(+), 1 deletion(-)

Rediffed for Debian's 2.4.27 Horms <horms@debian.org>

--- a/drivers/char/rocket.c	2005-08-02 16:45:34.000000000 +0900
+++ b/drivers/char/rocket.c	2005-08-02 16:45:58.000000000 +0900
@@ -257,7 +257,7 @@
 	printk("rp_do_receive(%d, %d)...", ToRecv, space);
 #endif
 	if (ToRecv == 0 || (space <= 0))
-		return;
+		goto done;
 	
 	/*
 	 * determine how many we can actually read in.  If we can't
@@ -352,6 +352,7 @@
 		count += ToRecv;
 	}
 	ld->receive_buf(tty, tty->flip.char_buf, tty->flip.flag_buf, count);
+done:
 	tty_ldisc_deref(ld);
 }
 
